File: module-zig/service-grpc/service_watch.zig
//! Mirrors module/service-grpc/service_watch.go

const std = @import("std");
const pb = @import("pb/etcdserverpb.pb.zig");

const errors = @import("error.zig");
const module = @import("module.zig");
const types = @import("types.zig");
const util = @import("util.zig");

// Backed by the wasm page allocator so it grows on demand (mirrors the Go
// module's GC), retaining capacity across per-message resets.
var arena_state = std.heap.ArenaAllocator.init(std.heap.wasm_allocator);
var out: [1536 * 1024]u8 = undefined;

fn arena() std.mem.Allocator {
    return arena_state.allocator();
}

fn decode(comptime T: type, b: []const u8) !T {
    var reader = std.Io.Reader.fixed(b);
    return T.decode(&reader, arena());
}

fn encode(msg: anytype) ![]const u8 {
    var writer = std.Io.Writer.fixed(&out);
    try msg.encode(&writer, arena());
    return writer.buffered();
}

pub fn shardRecv(name: []const u8, data: []const u8, id: u64) void {
    _ = name;
    if (id == types.WATCH_ID_ERROR) {
        std.debug.print("watch err {s}\n", .{data});
        return;
    }
    _ = arena_state.reset(.retain_capacity);
    switch (data[0]) {
        types.WatchMessageType_INIT => {
            const header = decode(pb.ResponseHeader, data[1..]) catch |err|
                std.debug.panic("Unable to unmarshal response header: {s}", .{@errorName(err)});
            const resp = pb.WatchResponse{
                .header = header,
                .watch_id = @bitCast(id),
                .created = true,
            };
            const b = encode(resp) catch |err|
                std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
            module.server.send(b) catch {};
        },
        types.WatchMessageType_EVENT => {
            const events = module.bufferPoolWatchEvent.find(id);
            if (events.append(data[1..])) {
                return;
            }
            var last_rev: u64 = 0;
            var resp = pb.WatchResponse{
                .header = .{},
                .watch_id = @bitCast(id),
            };
            var it = events.iterator();
            while (it.next()) |b| {
                last_rev = std.mem.readInt(u64, b[0..8], .big);
                const evt = decode(pb.Event, b[8..]) catch |err|
                    std.debug.panic("Unable to unmarshal event: {s}", .{@errorName(err)});
                resp.events.append(arena(), evt) catch |err|
                    std.debug.panic("{s}", .{@errorName(err)});
            }
            const current_rev = std.mem.readInt(u64, data[1..9], .big);
            if (last_rev == current_rev) {
                resp.fragment = true;
            }
            resp.header.?.revision = @bitCast(last_rev);
            const res = encode(resp) catch |err|
                std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
            module.server.send(res) catch {};
            events.reset();
            if (!events.append(data[1..])) {
                @panic("Failed to append watch event after reset");
            }
        },
        types.WatchMessageType_SYNC => {
            const events = module.bufferPoolWatchEvent.find(id);
            var resp = pb.WatchResponse{
                .watch_id = @bitCast(id),
            };
            resp.header = decode(pb.ResponseHeader, data[1..]) catch |err|
                std.debug.panic("Unable to unmarshal response header: {s}", .{@errorName(err)});
            var it = events.iterator();
            while (it.next()) |b| {
                const evt = decode(pb.Event, b[8..]) catch |err| {
                    std.debug.print("{d} {s}\n", .{ b.len, b });
                    events.reset();
                    std.debug.panic("Unable to unmarshal event in sync: {s}", .{@errorName(err)});
                };
                resp.events.append(arena(), evt) catch |err|
                    std.debug.panic("{s}", .{@errorName(err)});
            }
            const b = encode(resp) catch |err|
                std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
            module.server.send(b) catch {};
            events.reset();
        },
        types.WatchMessageType_NOTIFY => {
            const header = decode(pb.ResponseHeader, data[1..]) catch |err|
                std.debug.panic("Unable to unmarshal response header: {s}", .{@errorName(err)});
            const resp = pb.WatchResponse{
                .header = header,
                .watch_id = @bitCast(id),
            };
            const b = encode(resp) catch |err|
                std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
            module.server.send(b) catch {};
        },
        types.WatchMessageType_CANCELED => {
            const resp = pb.WatchResponse{
                .watch_id = @bitCast(id),
                .canceled = true,
            };
            const b = encode(resp) catch |err|
                std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
            module.server.send(b) catch {};
            // TODO: reset watch buffer pool
        },
        types.WatchMessageType_ERR_EXISTS => {
            const resp = pb.WatchResponse{
                .watch_id = -1,
                .created = true,
                .canceled = true,
                .cancel_reason = errors.err_watcher_duplicate_id,
            };
            const b = encode(resp) catch |err|
                std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
            module.server.send(b) catch {};
        },
        types.WatchMessageType_ERR_COMPACTED => {
            const header = decode(pb.ResponseHeader, data[1..]) catch |err|
                std.debug.panic("Unable to unmarshal response header: {s}", .{@errorName(err)});
            var resp = pb.WatchResponse{
                .header = header,
                .watch_id = @bitCast(id),
                .created = true,
            };
            var b = encode(resp) catch |err|
                std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
            module.server.send(b) catch {};
            resp = pb.WatchResponse{
                .header = header,
                .watch_id = @bitCast(id),
                .canceled = true,
                .compact_revision = header.revision,
            };
            b = encode(resp) catch |err|
                std.debug.panic("Unable to marshal watch response: {s}", .{@errorName(err)});
            module.server.send(b) catch {};
        },
        else => @panic("Unrecognized"),
    }
}

pub fn open() anyerror!void {
    return util.kvShard().StreamOpen("watch");
}

pub fn recv(data: []const u8) anyerror!void {
    return util.kvShard().StreamSend("watch", data);
}

pub fn close() anyerror!void {}

File: module/service-grpc/service_watch.go
package main

import (
	"encoding/binary"
	"errors"
	"sync"

	"github.com/pantopic/wazero-buffer-pool/sdk-go"
	"github.com/pantopic/wazero-grpc-server/sdk-go"

	internal "github.com/pantopic/turbokube/module/service-grpc/internal"
)

var evtPool = sync.Pool{New: func() any { return &internal.Event{} }}
var watchEventBatch = &internal.WatchEventBatch{Event: &internal.Event{}}
var watchEventSync = &internal.WatchEventSync{}

func shardRecv(_, data []byte, id uint64) {
	var err error
	if id == WATCH_ID_ERROR {
		println(`watch err ` + string(data))
		err = errors.New(string(data))
		return
	}
	switch data[0] {
	case WatchMessageType_INIT:
		watchResp.Reset()
		respHeader.Reset()
		if err = respHeader.UnmarshalVT(data[1:]); err != nil {
			panic(`Unable to unmarshal response header: ` + err.Error())
		}
		watchResp.Header = respHeader
		watchResp.WatchId = int64(id)
		watchResp.Created = true
		if data, err = watchResp.MarshalVT(); err != nil {
			panic(`Unable to marshal watch response: ` + err.Error())
		}
		grpc_server.Send(data)
	case WatchMessageType_EVENT_BATCH:
		watchEventBatch.Reset()
		if err = watchEventBatch.UnmarshalVT(data[1:]); err != nil {
			panic(`Unable to unmarshal watch event batch: ` + err.Error())
		}
		if len(watchEventBatch.WatchIdsPrev) > 0 {
			b, err := watchEventBatch.Event.MarshalVT()
			if err != nil {
				panic(`Unable to marshal watch event: ` + err.Error())
			}
			for _, id := range watchEventBatch.WatchIdsPrev {
				sendEvent(id, watchEventBatch.Revision, b, watchEventBatch.Event)
			}
		}
		if len(watchEventBatch.WatchIds) > 0 {
			watchEventBatch.Event.PrevKv = nil
			b, err := watchEventBatch.Event.MarshalVT()
			if err != nil {
				panic(`Unable to marshal watch event: ` + err.Error())
			}
			for _, id := range watchEventBatch.WatchIds {
				sendEvent(id, watchEventBatch.Revision, b, watchEventBatch.Event)
			}
		}
	case WatchMessageType_EVENT_SYNC:
		watchEventSync.Reset()
		if err = watchEventSync.UnmarshalVT(data[1:]); err != nil {
			panic(`Unable to unmarshal watch event batch: ` + err.Error())
		}
		for _, id := range watchEventSync.IDs {
			events := bufferPoolWatchEvent.Find(uint64(id))
			clearEvents(events, id, watchEventSync.Revision, nil, nil)
		}
	case WatchMessageType_NOTIFY:
		watchResp.Reset()
		respHeader.Reset()
		if err = respHeader.UnmarshalVT(data[1:]); err != nil {
			panic(`Unable to unmarshal response header: ` + err.Error())
		}
		watchResp.Header = respHeader
		watchResp.WatchId = int64(id)
		if data, err = watchResp.MarshalVT(); err != nil {
			panic(`Unable to marshal watch response: ` + err.Error())
		}
		grpc_server.Send(data)
	case WatchMessageType_CANCELED:
		watchResp.Reset()
		watchResp.WatchId = int64(id)
		watchResp.Canceled = true
		if data, err = watchResp.MarshalVT(); err != nil {
			panic(`Unable to marshal watch response: ` + err.Error())
		}
		grpc_server.Send(data)
		// TODO: reset watch buffer pool
	case WatchMessageType_ERR_EXISTS:
		watchResp.Reset()
		watchResp.WatchId = -1
		watchResp.Created = true
		watchResp.Canceled = true
		watchResp.CancelReason = ErrWatcherDuplicateID.Error()
		if data, err = watchResp.MarshalVT(); err != nil {
			panic(`Unable to marshal watch response: ` + err.Error())
		}
		grpc_server.Send(data)
	case WatchMessageType_ERR_COMPACTED:
		if err = respHeader.UnmarshalVT(data[1:]); err != nil {
			panic(`Unable to unmarshal response header: ` + err.Error())
		}
		watchResp.Reset()
		watchResp.Header = respHeader
		watchResp.WatchId = int64(id)
		watchResp.Created = true
		if data, err = watchResp.MarshalVT(); err != nil {
			panic(`Unable to marshal watch response: ` + err.Error())
		}
		grpc_server.Send(data)
		watchResp.Reset()
		watchResp.Header = respHeader
		watchResp.WatchId = int64(id)
		watchResp.Canceled = true
		watchResp.CompactRevision = respHeader.Revision
		if data, err = watchResp.MarshalVT(); err != nil {
			panic(`Unable to marshal watch response: ` + err.Error())
		}
		grpc_server.Send(data)
	default:
		panic(`Unrecognized`)
	}
}

func sendEvent(id int64, rev uint64, b []byte, evt *internal.Event) {
	events := bufferPoolWatchEvent.Find(uint64(id))
	b2 := binary.BigEndian.AppendUint64(b, rev)
	if events.Append(b2) {
		return
	}
	clearEvents(events, id, rev, b, evt)
	if !events.Append(b2) {
		panic(`Failed to append watch event after reset`)
	}
}

func clearEvents(events buffer_pool.MultiValue, id int64, rev uint64, b []byte, evt *internal.Event) {
	var lastRev uint64
	resp := &internal.WatchResponse{
		Header:  &internal.ResponseHeader{},
		WatchId: int64(id),
	}
	for b := range events.Iter() {
		evt := &internal.Event{}
		if err := evt.UnmarshalVT(b[:len(b)-8]); err != nil {
			panic(`Unable to unmarshal event B: ` + err.Error())
		}
		lastRev = binary.BigEndian.Uint64(b[len(b)-8:])
		resp.Events = append(resp.Events, evt)
	}
	if len(resp.Events) == 0 {
		return
	}
	if len(b) > 0 {
		var recycle bool
		if evt == nil {
			recycle = true
			evt := evtPool.Get().(*internal.Event)
			evt.Reset()
			if err := evt.UnmarshalVT(b); err != nil {
				panic(`Unable to unmarshal event C: ` + err.Error())
			}
		}
		if lastRev == rev {
			resp.Fragment = true
		}
		if recycle {
			defer evtPool.Put(evt)
		}
		resp.Header.Revision = int64(lastRev)
	} else {
		resp.Header.Revision = int64(rev)
	}
	res, err := resp.MarshalVT()
	if err != nil {
		panic(`Unable to marshal watch response: ` + err.Error())
	}
	grpc_server.Send(res)
	events.Reset()
}

var watchResp = &internal.WatchResponse{
	Header: &internal.ResponseHeader{},
}
var respHeader = &internal.ResponseHeader{}

func watchOpen() (err error) {
	return kvShard().StreamOpen([]byte(`watch`))
}

func watchRecv(data []byte) (err error) {
	return kvShard().StreamSend([]byte(`watch`), data)
}

func watchClose() (err error) {
	return
}
